### DISCUSSION TITLE
script/bootstrap does not work

### DISCUSSION THREAD

**Discussion by tohoyn** [atom/atom#23563] [Q&A]
I forked atom/atom and cloned it to a local repository. Command script/bootstrap gives the following error:
`/usr/bin/env: 'node': No such file or directory`
What should I do?`

**Comment by icecream17** [comment]
You need to install `node` and `python` first before running script/bootstrap

Make sure to add them to the path

The easiest way on windows (also great for updating) is to install chocolately, then run `choco install nodejs` and `choco install python`

Edit: see below

**Comment by tohoyn** [comment]
Now I get the following error:
***
tohoyn@th-bookworm:~/github/atom$ LANG=C script/bootstrap 
Node:	v17.4.0
Python:	v3.9.10
Installing script dependencies
npm ERR! code 1
npm ERR! path /home/tohoyn/github/atom/script/node_modules/fs-admin
npm ERR! command failed
npm ERR! command sh -c prebuild-install || node-gyp rebuild
npm ERR! make: Entering directory '/home/tohoyn/github/atom/script/node_modules/fs-admin/build'
npm ERR!   CXX(target) Release/obj.target/fs_admin/src/main.o
npm ERR! make: Leaving directory '/home/tohoyn/github/atom/script/node_modules/fs-admin/build'
npm ERR! In file included from ../src/main.cc:1:
npm ERR! ../../nan/nan.h: In function 'void Nan::AsyncQueueWorker(Nan::AsyncWorker*)':
npm ERR! ../../nan/nan.h:2298:7: warning: cast between incompatible function types from 'void (*)(uv_work_t*)' {aka 'void (*)(uv_work_s*)'} to 'uv_after_work_cb' {aka 'void (*)(uv_work_s*, int)'} [-Wcast-function-type]
npm ERR!  2298 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
npm ERR!       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
npm ERR! In file included from ../../nan/nan.h:2884,
npm ERR!                  from ../src/main.cc:1:
npm ERR! ../../nan/nan_typedarray_contents.h: In constructor 'Nan::TypedArrayContents<T>::TypedArrayContents(v8::Local<v8::Value>)':
npm ERR! ../../nan/nan_typedarray_contents.h:34:43: error: 'class v8::ArrayBuffer' has no member named 'GetContents'
npm ERR!    34 |       data   = static_cast<char*>(buffer->GetContents().Data()) + byte_offset;
npm ERR!       |                                           ^~~~~~~~~~~
npm ERR! make: *** [fs_admin.target.mk:114: Release/obj.target/fs_admin/src/main.o] Error 1
npm ERR! gyp ERR! build error 
npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
npm ERR! gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:520:28)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
npm ERR! gyp ERR! System Linux 5.15.0-3-amd64
npm ERR! gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/tohoyn/github/atom/script/node_modules/fs-admin
npm ERR! gyp ERR! node -v v17.4.0
npm ERR! gyp ERR! node-gyp -v v8.4.1
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/tohoyn/.npm/_logs/2022-02-05T12_32_34_659Z-debug-0.log
node:child_process:869
    throw err;
    ^

Error: Command failed: npm --loglevel=error install
npm ERR! code 1
npm ERR! path /home/tohoyn/github/atom/script/node_modules/fs-admin
npm ERR! command failed
npm ERR! command sh -c prebuild-install || node-gyp rebuild
npm ERR! make: Entering directory '/home/tohoyn/github/atom/script/node_modules/fs-admin/build'
npm ERR!   CXX(target) Release/obj.target/fs_admin/src/main.o
npm ERR! make: Leaving directory '/home/tohoyn/github/atom/script/node_modules/fs-admin/build'
npm ERR! In file included from ../src/main.cc:1:
npm ERR! ../../nan/nan.h: In function 'void Nan::AsyncQueueWorker(Nan::AsyncWorker*)':
npm ERR! ../../nan/nan.h:2298:7: warning: cast between incompatible function types from 'void (*)(uv_work_t*)' {aka 'void (*)(uv_work_s*)'} to 'uv_after_work_cb' {aka 'void (*)(uv_work_s*, int)'} [-Wcast-function-type]
npm ERR!  2298 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
npm ERR!       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
npm ERR! In file included from ../../nan/nan.h:2884,
npm ERR!                  from ../src/main.cc:1:
npm ERR! ../../nan/nan_typedarray_contents.h: In constructor 'Nan::TypedArrayContents<T>::TypedArrayContents(v8::Local<v8::Value>)':
npm ERR! ../../nan/nan_typedarray_contents.h:34:43: error: 'class v8::ArrayBuffer' has no member named 'GetContents'
npm ERR!    34 |       data   = static_cast<char*>(buffer->GetContents().Data()) + byte_offset;
npm ERR!       |                                           ^~~~~~~~~~~
npm ERR! make: *** [fs_admin.target.mk:114: Release/obj.target/fs_admin/src/main.o] Error 1
npm ERR! gyp ERR! build error 
npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
npm ERR! gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:520:28)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
npm ERR! gyp ERR! System Linux 5.15.0-3-amd64
npm ERR! gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/tohoyn/github/atom/script/node_modules/fs-admin
npm ERR! gyp ERR! node -v v17.4.0
npm ERR! gyp ERR! node-gyp -v v8.4.1
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/tohoyn/.npm/_logs/2022-02-05T12_32_34_659Z-debug-0.log

    at checkExecSyncError (node:child_process:828:11)
    at Object.execFileSync (node:child_process:866:15)
    at module.exports (/home/tohoyn/github/atom/script/lib/install-script-dependencies.js:13:16)
    at Object.<anonymous> (/home/tohoyn/github/atom/script/bootstrap:37:1)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  status: 1,
  signal: null,
  output: [
    null,
    Buffer(0) [Uint8Array] [],
    Buffer(2400) [Uint8Array] [
      110, 112, 109,  32,  69,  82,  82,  33,  32,  99, 111, 100,
      101,  32,  49,  10, 110, 112, 109,  32,  69,  82,  82,  33,
       32, 112,  97, 116, 104,  32,  47, 104, 111, 109, 101,  47,
      116, 111, 104, 111, 121, 110,  47, 103, 105, 116, 104, 117,
       98,  47,  97, 116, 111, 109,  47, 115,  99, 114, 105, 112,
      116,  47, 110, 111, 100, 101,  95, 109, 111, 100, 117, 108,
      101, 115,  47, 102, 115,  45,  97, 100, 109, 105, 110,  10,
      110, 112, 109,  32,  69,  82,  82,  33,  32,  99, 111, 109,
      109,  97, 110, 100,
      ... 2300 more items
    ]
  ],
  pid: 5954,
  stdout: Buffer(0) [Uint8Array] [],
  stderr: Buffer(2400) [Uint8Array] [
    110, 112, 109,  32,  69,  82,  82,  33,  32,  99, 111, 100,
    101,  32,  49,  10, 110, 112, 109,  32,  69,  82,  82,  33,
     32, 112,  97, 116, 104,  32,  47, 104, 111, 109, 101,  47,
    116, 111, 104, 111, 121, 110,  47, 103, 105, 116, 104, 117,
     98,  47,  97, 116, 111, 109,  47, 115,  99, 114, 105, 112,
    116,  47, 110, 111, 100, 101,  95, 109, 111, 100, 117, 108,
    101, 115,  47, 102, 115,  45,  97, 100, 109, 105, 110,  10,
    110, 112, 109,  32,  69,  82,  82,  33,  32,  99, 111, 109,
    109,  97, 110, 100,
    ... 2300 more items
  ]
}

Node.js v17.4.0
tohoyn@th-bookworm:~/github/atom$ 

***

**Reply by icecream17** [reply]
Seems like this error would be fixed by https://github.com/nodejs/nan/pull/888

Not sure why it isn't fixed

**Reply by tohoyn** [reply]
How do I get Atom script/boostrap to use the fixed version of nan? I tried to fork nan, pull #888, and install the fixed nan with npm install --save nan, but I get the same error.

**Reply by icecream17** [reply]
Edit 2: So far I've found that newer versions of node fail bc `GetContents` (in v8?) doesn't exist anymore, so install `node@14` (which includes `npm@6`)

Although this is a very suboptimal solution. [The fs-admin repository](https://github.com/atom/fs-admin) is now updated, so a new version needs to be used by atom.

**Reply by tohoyn** [reply]
I finally managed to get the script to use the fixed version of nan. But the atom command does not work.

**Reply by icecream17** [reply]
Might need to add \<the folder where the atom binary is> to the "Path"

https://github.com/icecream17/atom-update-backlog/blob/main/Debugging-atom.md#adding-atom-to-path--adding-atom-shell-commands
